lib/puppet/type/service.rb

```diff
@@ -81,11 +81,18 @@ def retrieve
       end
 
       # This only makes sense on systemd systems. Static services cannot be enabled
-      # or disabled manually.
+      # or disabled manually. Indirect services should not be enabled or disabled due 
+      # to limitations in systemd (see https://github.com/systemd/systemd/issues/6681)
       def insync?(current)
-        if provider.respond_to?(:cached_enabled?) && provider.cached_enabled? == 'static'
-          Puppet.debug("Unable to enable or disable static service #{@resource[:name]}")
-          return true
+        if provider.respond_to?(:cached_enabled?)
+          case provider.cached_enabled?
+          when 'static'
+            Puppet.debug("Unable to enable or disable static service #{@resource[:name]}")
+            return true
+          when 'indirect'
+            Puppet.warning("Service #{@resource[:name]} is in 'indirect' state and cannot be enabled/disabled")
+            return true
+          end
```

Comment:
We are at a very top level here and we're taking decisions for all platforms without having knowledge about their concept of service. Could we move this logic to `systemd` provider and let and have something like here:
```ruby

def insync?(current)
  provider.insync?(current) if provider.respond_to(:insync?)
  super(current)
end

This way each provider could have his own `insync?` based on his understanding and particularities of `service`

Reply:
Something similar is used for group members: 
https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/group.rb#L97-L103

I'm not sure if `provider.insync?` is not already used to sync entire resource, if present on provider. You could use `enabled_insync?` if this is the case.

Reply:
Went with `enabled_insync?` since `insync?` methods are usually per properties/fields.